server/mailet/rate-limiter-redis/src/main/scala/org/apache/james/rate/limiter/redis/RedisRateLimiter.scala

```diff
@@ -0,0 +1,51 @@
+/****************************************************************
+ * Licensed to the Apache Software Foundation (ASF) under one   *
+ * or more contributor license agreements.  See the NOTICE file *
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ * http://www.apache.org/licenses/LICENSE-2.0                   *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.rate.limiter.redis
+
+import es.moki.ratelimitj.core.limiter.request.{ReactiveRequestRateLimiter, RequestLimitRule}
+import es.moki.ratelimitj.redis.request.{RedisRateLimiterFactory => RateLimitjFactory}
+import io.lettuce.core.RedisClient
+import org.apache.james.rate.limiter.api.Quantity.Quantity
+import org.apache.james.rate.limiter.api.{AcceptableRate, RateExceeded, RateLimiter, RateLimiterFactory, RateLimitingKey, RateLimitingResult, Rule, Rules}
+import org.reactivestreams.Publisher
+import reactor.core.scala.publisher.SMono
+
+import scala.jdk.CollectionConverters._
+
+class RedisRateLimiterFactory(redisConfiguration: RedisRateLimiterConfiguration) extends RateLimiterFactory {
+  val rateLimitjFactory: RateLimitjFactory = new RateLimitjFactory(RedisClient.create(redisConfiguration.redisURI))
```

Comment:
Stupid question: any way to connect to a redis cluster?

Reply:
Having a look at https://github.com/lettuce-io/lettuce-core/blob/6.1.6.RELEASE/src/test/java/io/lettuce/examples/ConnectToRedisCluster.java we might want a `boolean enableClustering()` parameter. If true we likely should use `RedisClusterClient`. What do you think?

Reply:
- Ratelimitj lib also support `RedisClusterClient`, (`es.moki.ratelimitj.redis.request.RedisClusterRateLimiterFactory`)
- We can re-defined the mailet in XML files.
Ex: 
```xml
<redisUris>url1;url2;url3</redisUris>
```

// My view: I don't think "Redis cluster" is a "main" of the rate-limit feature. 
- "Storage size": We have a TTL parameter when setting data to Redis, then data will auto removed after expired. It won't take up much memory. 
- "IO throughput": it will not problem with In-memory DB? (I'm not sure)
- "Persistence": as written in ADR, this is a low priority.

Reply:
> // My view: I don't think "Redis cluster" is a "main" of the rate-limit feature.

IMO it is a core feature: one single Redis node and we can't really achieve **high availability** / **scalability**!

IMO we should provide a way for people to connect to a Redis cluster.

However I would agree to do it in another PR.